ikarus/finiteelements/mechanics/materials/hyperelastic/deviatoric/interface.hh

```diff
@@ -0,0 +1,159 @@
+// SPDX-FileCopyrightText: 2021-2024 The Ikarus Developers mueller@ibb.uni-stuttgart.de
+// SPDX-License-Identifier: LGPL-3.0-or-later
+
+/**
+ * \file interface.hh
+ * \brief Implementation of the interface for the deviatoric part of a hyperelastic material.
+ * \ingroup  materials
+ */
+
+#pragma once
+
+#include <dune/common/float_cmp.hh>
+
+#include <ikarus/finiteelements/mechanics/materials/hyperelastic/concepts.hh>
+#include <ikarus/utils/tensorutils.hh>
+
+namespace Ikarus::Materials {
+
+/**
+ * \brief This is the interface implementation for the deviatoric part of a hyperelastic material.
+ *    It is intended to be used with the hyperelastic material model.
+ * \ingroup materials
+ *
+ * \details The deviatoric part of the hyperelastic model, i.e., related to
+ * \f$ \hat{\Psi}(\la_1, \la_2, \la_3) \f$, is parametrized with a certain
+ * deviatoric function (DF) implemented in terms of principal stretches.
+ * The three interface functions (energy, streses and tangentModulus) are called with the argument being the principal
+ * stretches (\f$ \la_i \f$). The underlying deviatoric function must only implement the energy
+ * \f$ \hat{\Psi}(\la_1, \la_2, \la_3) \f$ and its first and second derivatives
+ * w.r.t the total principal stretches.
+ *
+ * \tparam DF Deviatoric function.
+ */
+template <Concepts::DeviatoricFunction DF>
+struct Deviatoric
+{
+  using ScalarType         = typename DF::ScalarType;
+  using PrincipalStretches = typename DF::PrincipalStretches;
+  using MaterialParameters = typename DF::MaterialParameters;
+
+  using DeviatoricFunction = DF;
+
+  using FirstDerivative  = typename DF::FirstDerivative;
+  using SecondDerivative = typename DF::SecondDerivative;
+
+  static constexpr int dim = 3;
+
+  using StressMatrix   = Eigen::Vector<ScalarType, dim>;
+  using MaterialTensor = Eigen::TensorFixedSize<ScalarType, Eigen::Sizes<3, 3, 3, 3>>;
+
+  [[nodiscard]] constexpr static std::string name() noexcept { return "Deviatoric function: " + DF::name(); }
+
+  Deviatoric(const DF df)
+      : deviatoricFunction_{df} {}
+
+  /**
+   * \brief Returns the material parameters stored in the deviatoric part of the material.
+   */
+  const MaterialParameters materialParameters() const { return deviatoricFunction_.materialParametersImpl(); }
+
+  /**
+   * \brief Returns the stored energy obtained from the deviatoric function.
+   *
+   * \param lambdas the principal stretches.
+   * \return ScalarType the energy.
+   */
+  ScalarType storedEnergy(const PrincipalStretches& lambda) const {
+    checkDuplicates(lambda);
+    return deviatoricFunction_.storedEnergyImpl(lambda);
+  };
+
+  /**
+   * \brief Returns the principal PK2 stresses obtained from the first derivative of the deviatoric function.
+   *
+   * \param lambda the principal stretches.
+   * \return StressMatrix the stresses in Cartesian coordinate system.
```

Comment:
These returned stresses are only the eigenvalues of S arent they? The comment is wrong?

Reply:
Yeah I think the comment is wrong here (also in matTangent). In my understanding these are the stresses in prinicpal coordinate system and not the stress in cartesian coordinate system